extension/apikeyauthextension/authenticator_test.go

```diff
@@ -0,0 +1,336 @@
+
+import (
+	"context"
+	"encoding/base64"
+	"encoding/json"
+	"fmt"
+	"net/http"
+	"net/http/httptest"
+	"testing"
+
+	"github.com/elastic/go-elasticsearch/v8/typedapi/security/hasprivileges"
+	"github.com/elastic/go-elasticsearch/v8/typedapi/types"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+	"go.opentelemetry.io/collector/client"
+	"go.opentelemetry.io/collector/component/componenttest"
+)
+
+const user = "test"
+
+var successfulResponse = hasprivileges.Response{
+	Application: map[string]types.ResourcePrivileges{
+		"my_app": map[string]types.Privileges{
+			"the_resource": {
+				"write": true,
+			},
+		},
+	},
+	Username:        user,
+	HasAllRequested: true,
+}
+
+func TestAuthenticator(t *testing.T) {
+	for name, testcase := range map[string]struct {
+		handler http.HandlerFunc
+		header  map[string][]string
+
+		expectedUsername string
+		expectedAPIKeyID string
+		expectedErr      string
+	}{
+		"success": {
+			handler:          newCannedHasPrivilegesHandler(successfulResponse),
+			expectedUsername: "test",
+			expectedAPIKeyID: "id",
+		},
+		"error": {
+			handler: newCannedErrorHandler(types.ElasticsearchError{
+				ErrorCause: types.ErrorCause{
+					Type: "a_type",
+					Reason: func() *string {
+						reason := "a_reason"
+						return &reason
+					}(),
+				},
+				Status: 400,
+			}),
+			expectedErr: `status: 400, failed: [a_type], reason: a_reason`,
+		},
+		"missing_privileges": {
+			handler:     newCannedHasPrivilegesHandler(hasprivileges.Response{HasAllRequested: false}),
+			expectedErr: `API Key "id" unauthorized`,
+		},
+	} {
+		t.Run(name, func(t *testing.T) {
+			srv := newMockElasticsearch(t, testcase.handler)
+			authenticator := newTestAuthenticator(t, srv, createDefaultConfig().(*Config))
+
+			ctx, err := authenticator.Authenticate(context.Background(), map[string][]string{
+				"Authorization": {"ApiKey " + base64.StdEncoding.EncodeToString([]byte("id:secret"))},
+			})
+			if testcase.expectedErr != "" {
+				assert.EqualError(t, err, testcase.expectedErr)
+				return
+			}
+			assert.NoError(t, err)
+
+			clientInfo := client.FromContext(ctx)
+			require.NotNil(t, clientInfo.Auth)
+			assert.Equal(t, []string{"username", "api_key"}, clientInfo.Auth.GetAttributeNames())
+			assert.Equal(t, testcase.expectedUsername, clientInfo.Auth.GetAttribute("username"))
+			assert.Equal(t, testcase.expectedAPIKeyID, clientInfo.Auth.GetAttribute("api_key"))
+		})
+	}
+}
+
+func TestAuthenticator_ApplicationPrivileges(t *testing.T) {
+	srv := newMockElasticsearch(t, func(w http.ResponseWriter, r *http.Request) {
+		var body hasprivileges.Request
+		assert.NoError(t, json.NewDecoder(r.Body).Decode(&body))
+		assert.Equal(t, hasprivileges.Request{
+			Application: []types.ApplicationPrivilegesCheck{{
+				Application: "my_app",
+				Resources:   []string{"the_resource"},
+				Privileges:  []string{"read", "write"},
+			}},
+		}, body)
+		_ = json.NewEncoder(w).Encode(successfulResponse)
```

Comment:
```suggestion
		assert.NoError(t, json.NewEncoder(w).Encode(successfulResponse))
```

 Should we assert on the response encoding?